core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -36,373 +84,287 @@ class Queue[A] private (
-                if (surplusValues.isEmpty)
-                  p.complete(true) -> Surplus(
-                    IQueue.empty.enqueue(addToQueue.toList),
-                    IQueue.empty,
-                    hook
-                  )
-                else
-                  strategy match {
-                    case BackPressure =>
-                      IO.now(true) -> Surplus(
-                        IQueue.empty[A].enqueue(addToQueue.toList),
-                        IQueue.empty.enqueue(surplusValues -> p),
-                        hook
-                      )
-                    case Sliding =>
-                      val notTaken = addToQueue ++ surplusValues
-                      val toQueue  = capacity.fold(notTaken)(notTaken.takeRight)
-                      p.complete(false) -> Surplus(
-                        IQueue.empty.enqueue(toQueue.toList),
-                        IQueue.empty,
-                        hook
-                      )
-                    case Dropping =>
-                      p.complete(false) -> Surplus(
-                        IQueue.empty.enqueue(addToQueue.toList),
-                        IQueue.empty,
-                        hook
-                      )
-                  }
-
-              completeTakers *> complete -> surplus
-            } else
-              completeTakers *> p.complete(true) -> Deficit(deficitValues, hook)
-        }
-
-      case (p, Surplus(values, putters, hook)) =>
-        val (addToQueue, surplusValues) =
-          capacity.fold((as, Iterable.empty[A]))(c => as.splitAt(c - values.size))
-        if (surplusValues.isEmpty)
-          (p.complete(true), Surplus(values.enqueue(addToQueue.toList), putters, hook))
-        else {
-          strategy match {
-            case BackPressure =>
-              (
-                IO.now(true),
-                Surplus(values.enqueue(addToQueue.toList), putters.enqueue(surplusValues -> p), hook)
-              )
-            case Sliding =>
-              (
-                p.complete(true),
-                capacity.fold(
-                  Surplus(values.enqueue(as.toList), putters, hook)
-                ) { c =>
-                  Surplus(
-                    values
-                      .takeRight(c - as.size)
-                      .enqueue(as.takeRight(c).toList),
-                    putters,
-                    hook
-                  )
-                }
-              )
-            case Dropping =>
-              (
-                p.complete(false),
-                Surplus(values.enqueue(addToQueue.toList), putters, hook)
-              )
-          }
-        }
-
-      case (p, state @ Shutdown) => (p.interrupt, state)
-    }
-
-    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
-      case (_, p) => p.poll.void <> removePutter(p)
-    }
-
-    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
-  }
+  final def offerAll(as: Iterable[A]): IO[Nothing, Boolean] =
```

Comment:
Seeing as how this is backed by the ring buffer now, I'm tempted to make `offer(a: A)` the primitive operation, so it will be fastest, and implement `offerAll` in the simplest way: `IO.traverse(as)(offer(_))`. A ring buffer is not ideally suited to "batch upload" but is fast enough it's probably not necessary to specialize that.

Reply:
I initially did that but I changed to the current code for 2 reasons:
- it's faster to do the batch offer in a single `unsafe` method rather than going through an IO loop for each item
- in BackPressure strategy, we're enqueuing all the items in `putters` at once. If we implement `offerAll` using `offer`, we will only enqueue the first item then wait until it's processed, then we will enqueue the second item, etc. It means multiple calls to `offerAll` will not be processed in order.

Reply:
Sounds good. 👍